Version 1.5

GeorgeGkas opened the pull request:
We update the software to version 1.5
---

### New simplified macros
Now we provide a cleaner option to generate errors without the need to provide the flag parameter. 

We defined macros based on the warning flags.

- `slog_none()`
- `slog_live()`
- `slog_info()`
- `slog_warn()`
- `slog_debug()`
- `slog_error()`
- `slog_fatal()`
- `slog_panic()`

Each macro take the following parameters:

 1. The log level.
 2. A formated string. Format tags prototype follows the same rules as the C standard library function `printf()`.
 3. Additional arguments. There should be the same number of these arguments as the number of `%-tags` that expect a value.

Bellow we provide an example that logs a debug message:

```c
slog_debug(0, "The %s contains between %d and %d billion stars and at least %d billion planets.  ", "Milky Way", 200, 400, 100);
```

In addition, we added the option to print the corresponding file name and line number where a slog macro was called. This rule follows the macros which relate to a critical flag, and shown bellow:

- `slog_warn()`
- `slog_error()`
- `slog_fatal()`
- `slog_panic()`

Basic example:
```c
/* Log and print something fatal. */
slog_fatal(0, "Fatal message. We fell into dead zone.");
```
With expected output to be:

    2017.01.22-19:03:17.03 - [FATAL] <example.c:71> -- Fatal message. We fell into dead zone.

### Updated Readme
We further updated the readme file. Now we provide better examples and explanations. Please refer to `Readme.md` file to read more.

### Updated Examples
We included more examples.

Please review this version. Thank you all.

GeorgeGks committed:
Code style cleanup, updated outdated examples.
We also added better funtion descriptions in header file.

GeorgeGks committed:
Updated to version 1.5.
We now can support simplified logging without the need to add flags. We also added a feature that allows the developer
to see in what line of her source file the corresponding error was throw. This adds better flexibility for project maintainers.
In the end, we further updated the Readme file and added better explanations.

GeorgeGks committed:
Fixed table in Readme file.

GeorgeGks committed:
Changed false image in Readme file.

Burnett01 commented on src/slog.h:
```diff
@@ -27,44 +28,89 @@
-#endif //DARWIN
-
-/* Loging flags */
-#define SLOG_NONE   0
```
I feel we should additionally keep these flags, in order not to break backward-compatibility. I've seen some closed-source projects that use these flags.

Proposed change:
```c
/* Loging flags. */
#define SLOG_FNONE   0
#define SLOG_FLIVE   1
#define SLOG_FINFO   2
#define SLOG_FWARN   3
#define SLOG_FDEBUG  4
#define SLOG_FERROR  5
#define SLOG_FFATAL  6
#define SLOG_FPANIC  7

/* Compat. Logging flags. */
#define SLOG_NONE   0
#define SLOG_LIVE   1
#define SLOG_INFO   2
#define SLOG_WARN   3
#define SLOG_DEBUG  4
#define SLOG_ERROR  5
#define SLOG_FATAL  6
#define SLOG_PANIC  7
````

Burnett01 commented on README.md:
```diff
@@ -42,50 +49,100 @@ PRETTYLOG 0
 ### Logging flags
 Slog has its logging flags to print something with status code.
 
-- SLOG_LIVE
```
Let's additionally keep up with legacy flags. Please check my comments above.

Maybe add something along:

### Logging flags
 Slog has its logging flags to print something with status code.

- `SLOG_FLIVE`
- `SLOG_FINFO`
- `SLOG_FWARN`
- `SLOG_FDEBUG`
- `SLOG_FERROR`
- `SLOG_FFATAL`
- `SLOG_FPANIC`
- `SLOG_FNONE`

*Please note:* As of version 1.5 the flags have changed. In order to keep backward-compatibility, the following legacy flags can be used as usual:

- `SLOG_LIVE`
- `SLOG_INFO`
- `SLOG_WARN`
- `SLOG_DEBUG`
- `SLOG_ERROR`
- `SLOG_FATAL`
- `SLOG_PANIC`
- `SLOG_NONE`

Burnett01 commented on README.md:
```diff
@@ -42,50 +49,100 @@ PRETTYLOG 0
 
 ### Print and log something
-Here is an example how use slog.
+Here is an example on how use slog.
```
"on how to use" :-)

Burnett01 commented on src/slog.c:
```diff
@@ -71,347 +81,352 @@ void slog_get_date(SlogDate *sdate)
-        /* Handle flags */
-        switch(flag) 
+        /* Handle flags. */
+        switch (flag)
```
In order to keep backward-compatibility, let's use this:

```c
/* Handle flags. */
switch (flag)
{
    case SLOG_LIVE:
    case SLOG_FLIVE:
        strncpy(color, CLR_NORMAL, sizeof(color));
        strncpy(alarm, "LIVE", sizeof(alarm));
        break;
    case SLOG_INFO:
    case SLOG_FINFO:
        strncpy(color, CLR_GREEN, sizeof(color));
        strncpy(alarm, "INFO", sizeof(alarm));
        break;
    case SLOG_WARN:
    case SLOG_FWARN:
        strncpy(color, CLR_YELLOW, sizeof(color));
        strncpy(alarm, "WARN", sizeof(alarm));
        break;
    case SLOG_DEBUG:
    case SLOG_FDEBUG:
        strncpy(color, CLR_BLUE, sizeof(color));
        strncpy(alarm, "DEBUG", sizeof(alarm));
        break;
    case SLOG_ERROR:
    case SLOG_FERROR:
        strncpy(color, CLR_RED, sizeof(color));
        strncpy(alarm, "ERROR", sizeof(alarm));
        break;
    case SLOG_FATAL:
    case SLOG_FFATAL:
        strncpy(color, CLR_RED, sizeof(color));
        strncpy(alarm, "FATAL", sizeof(alarm));
        break;
    case SLOG_PANIC:
    case SLOG_FPANIC:
        strncpy(color, CLR_RED, sizeof(color));
        strncpy(alarm, "PANIC", sizeof(alarm));
        break;
    case SLOG_NONE:
    case SLOG_FNONE:
    default:
        strncpy(prints, string, sizeof(string));
        flag = SLOG_FNONE;
        break;
}
```

Burnett01 replied:
As you can see above, ``SLOG_NONE`` and ``SLOG_FNONE`` both behave like ``default`` so we don't need the code twice (Dry) and save an additional if-clause (line [341](https://github.com/kala13x/slog/pull/12/files/1ea113f80999d38e13b984744ac6c2a172dfeed7#diff-0b7de98ff3b11d118c9550e81a75f340R341) and [360](https://github.com/kala13x/slog/pull/12/files/1ea113f80999d38e13b984744ac6c2a172dfeed7#diff-0b7de98ff3b11d118c9550e81a75f340R360))

Burnett01 reviewed (commented):
Yet again - great work dude! In order to keep backward-compatibility, these changes should be taken into consideration.

GeorgeGks committed:
We changed the flag names back to previous standards.
This update was proposed by @Burnett01 in the review of the #12 pull request,
and let us keep the backward compatibilty of our program.

GeorgeGkas commented:
Thank you very much @Burnett01 for your review. Backward compatibility is something that I have not consider. For the history part, I changed the flags because in the first test version of this update, I set the macro names the same as the flags (`SLOG_NONE`, `SLOG_FATAL`, etc), so I had to make this change. In the final revision I thought it was best to shift to lowercase names, but I forgot to change the flags back. I think we don't have to add more `case`s neither. Please, post review my last commit.

Burnett01 commented:
Great! Good work @GeorgeGkas

Burnett01 commented:
Hey @kala13x You up to get this merged? :) :+1:

seachdamh commented:
@GeorgeGkas Great work.

I have also issue.

 -> I need use logging library completly without config file. Still program have to configure things like enable write to file not only at start but also - change parameters in "working time" (like lvl of logged messages). Can I make it with present version without forking / changing source code of library?

GeorgeGkas commented:
@seachdamh This version I propose (1.5) keeps backward compatibility. If your are able to do this in the current stable release, then you are able to do this as well with 1.5. If there is no such option then you can propose an enchantment in issues page. For any further discussion, please open a new issue.
